src/runtime/containerd-shim-v2/create.go

```diff
@@ -259,3 +272,99 @@ func doMount(mounts []*containerd_types.Mount, rootfs string) error {
 	}
 	return nil
 }
+
+func configureNonRootHypervisor(runtimeConfig *oci.RuntimeConfig) error {
+	userName, err := createVmmUser()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			removeVmmUser(userName)
+		}
+	}()
+
+	u, err := user.Lookup(userName)
+	if err != nil {
+		return err
+	}
+
+	uid, err := strconv.Atoi(u.Uid)
+	if err != nil {
+		return err
+	}
+	gid, err := strconv.Atoi(u.Gid)
+	if err != nil {
+		return err
+	}
+	runtimeConfig.HypervisorConfig.Uid = uint32(uid)
+	runtimeConfig.HypervisorConfig.Gid = uint32(gid)
+
+	userTmpDir := path.Join("/run/user/", fmt.Sprint(uid))
+	dir, err := os.Stat(userTmpDir)
+	if os.IsNotExist(err) {
+		if err = os.Mkdir(userTmpDir, vc.DirMode); err != nil {
+			return err
+		}
+		if err = syscall.Chown(userTmpDir, uid, gid); err != nil {
+			return err
+		}
+	}
+	if dir != nil && !dir.IsDir() {
+		return fmt.Errorf("%s is expected to be a directory", userTmpDir)
+	}
+
+	if err := os.Setenv("XDG_RUNTIME_DIR", userTmpDir); err != nil {
+		return err
+	}
+
+	info, _ := os.Stat("/dev/kvm")
+	var Gid uint32
+	if stat, ok := info.Sys().(*syscall.Stat_t); ok {
+		Gid = stat.Gid
+	} else {
+		return fmt.Errorf("failed to get the gid of /dev/kvm")
+	}
+	// Add the kvm group to the hypervisor supplemental group so that the hypervisor process can access /dev/kvm
+	runtimeConfig.HypervisorConfig.Groups = append(runtimeConfig.HypervisorConfig.Groups, Gid)
+
+	return nil
+}
+
+func createVmmUser() (string, error) {
+	var (
+		err      error
+		userName string
+	)
+	maxAttempt := 5
+	attempt := 0
+
+	rbashPath := "/bin/rbash"
+	if _, err := os.Stat(rbashPath); os.IsNotExist(err) {
+		if _, err = utils.RunCommand([]string{"ln", "/bin/bash", rbashPath}); err != nil {
```

Comment:
Correct me if I'm misinterpreting, but, I don't think the runtime should be creating this link on the host, right?

Reply:
I added this to address @devimc 's comment on using `rbash` and create it if it doesn't exist. What's the concern of creating rbash?

Reply:
Do you really need this? Can you not great user without providing the login shell?

Reply:
@fengwang666 can you use  `/sbin/nologin` ?